File: ganilib-core/src/main/java/com/gani/lib/collection/SelfTruncatingSet.java
package com.gani.lib.collection;

import com.gani.lib.json.GJson;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.lang.reflect.Type;
import java.util.AbstractSet;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashSet;

public class SelfTruncatingSet<E> extends AbstractSet<E> implements Serializable {
  private static final long serialVersionUID = 1L;
  private static final Object VALUE = new Object();

  private SelfTruncatingMap<E, Object> backingMap;

  public SelfTruncatingSet(int maxCapacity) {
    this.backingMap = new SelfTruncatingMap<E, Object>(maxCapacity);
  }

  private int getMaxCapacity() {
    return backingMap.getMaxCapacity();
  }


  @Override
  public boolean add(E object) {
    return backingMap.put(object, VALUE) == VALUE;
  }

  @Override
  public void clear() {
    backingMap.clear();
  }

  @Override
  @SuppressWarnings("unchecked")
  public Object clone() {
    // If one day we need to support this, see HashSet implementation.
    throw new UnsupportedOperationException();
  }

  @Override
  public boolean contains(Object object) {
      return backingMap.containsKey(object);
  }

  @Override
  public boolean isEmpty() {
      return backingMap.isEmpty();
  }

  @Override
  public Iterator<E> iterator() {
      return backingMap.keySet().iterator();
  }

  @Override
  public boolean remove(Object object) {
      return backingMap.remove(object) != null;
  }

  @Override
  public int size() {
      return backingMap.size();
  }

  private void writeObject(ObjectOutputStream stream) throws IOException {
    // If one day we need to support this, see HashSet implementation.
    throw new UnsupportedOperationException();
  }

  private void readObject(ObjectInputStream stream) throws IOException, ClassNotFoundException {
    // If one day we need to support this, see HashSet implementation.
    throw new UnsupportedOperationException();
  }

  HashMap<E, HashSet<E>> createBackingMap(int capacity, float loadFactor) {
    return new HashMap<E, HashSet<E>>(capacity, loadFactor);
  }



  public static class GsonSerializer
      implements JsonSerializer<SelfTruncatingSet<?>>, JsonDeserializer<SelfTruncatingSet<?>> {
    @Override
    public JsonElement serialize(SelfTruncatingSet<?> src, Type typeOfSrc, JsonSerializationContext context) {
      JsonObject object = new JsonObject();
      object.addProperty("max_capacity", src.getMaxCapacity());
      object.addProperty("delegate", GJson.defaultGson().toJson(src));
      return object;
    }

    @Override
    public SelfTruncatingSet<?> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
      JsonObject jsonObject = json.getAsJsonObject();
      int maxCapacity = jsonObject.get("max_capacity").getAsInt();
      String delegate = jsonObject.get("delegate").getAsString();

      SelfTruncatingSet<Object> set = new SelfTruncatingSet<>(maxCapacity);
      // Use typeOfT so that the generic type is still maintained, e.g. LinkedHashSet<Long>
      LinkedHashSet<?> entries = GJson.defaultGson().fromJson(delegate, typeOfT);
      for (Object entry : entries) {
        set.add(entry);
      }
      return set;
    }
  }
}

File: glib-core/src/main/java/com/glib/core/collection/SelfTruncatingSet.kt
package com.glib.core.collection

import com.glib.core.json.GGson
import com.google.gson.*
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable
import java.lang.reflect.Type
import java.util.*

class SelfTruncatingSet<E>(maxCapacity: Int) : AbstractSet<E>(), Serializable {

    private val backingMap: SelfTruncatingMap<E, Any>

    private val maxCapacity: Int
        get() = backingMap.maxCapacity

    init {
        this.backingMap = SelfTruncatingMap(maxCapacity)
    }


    override fun add(`object`: E): Boolean {
        return backingMap.put(`object`, VALUE) === VALUE
    }

    override fun clear() {
        backingMap.clear()
    }

//    override fun clone(): Any {
//        // If one day we need to support this, see HashSet implementation.
//        throw UnsupportedOperationException()
//    }

//    override operator fun contains(`object`: Any): Boolean {
//        return backingMap.containsKey(`object`)
//    }

    override fun contains(element: E): Boolean {
        return backingMap.containsKey(element)
    }

    override fun isEmpty(): Boolean {
        return backingMap.isEmpty()
    }

    override fun iterator(): MutableIterator<E> {
        return backingMap.keys.iterator()
    }

    override fun remove(element: E): Boolean {
        return super.remove(element)
    }

//    override fun remove(`object`: Any): Boolean {
//        return backingMap.remove(`object`) != null
//    }

    override val size: Int
        get() = backingMap.size

//    override fun size(): Int {
//        return backingMap.size
//    }

    @Throws(IOException::class)
    private fun writeObject(stream: ObjectOutputStream) {
        // If one day we need to support this, see HashSet implementation.
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class, ClassNotFoundException::class)
    private fun readObject(stream: ObjectInputStream) {
        // If one day we need to support this, see HashSet implementation.
        throw UnsupportedOperationException()
    }

    internal fun createBackingMap(capacity: Int, loadFactor: Float): HashMap<E, HashSet<E>> {
        return HashMap(capacity, loadFactor)
    }


    class GsonSerializer : JsonSerializer<SelfTruncatingSet<*>>, JsonDeserializer<SelfTruncatingSet<*>> {
        override fun serialize(src: SelfTruncatingSet<*>, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
            val `object` = JsonObject()
            `object`.addProperty("max_capacity", src.maxCapacity)
            `object`.addProperty("delegate", GGson.defaultGson().toJson(src))
            return `object`
        }

        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): SelfTruncatingSet<*> {
            val jsonObject = json.asJsonObject
            val maxCapacity = jsonObject.get("max_capacity").asInt
            val delegate = jsonObject.get("delegate").asString

            val set = SelfTruncatingSet<Any>(maxCapacity)
            // Use typeOfT so that the generic type is still maintained, e.g. LinkedHashSet<Long>
            val entries = GGson.defaultGson().fromJson<LinkedHashSet<*>>(delegate, typeOfT)
            for (entry in entries) {
                set.add(entry)
            }
            return set
        }
    }

    companion object {
        private const val serialVersionUID = 1L
        private val VALUE = Any()
    }
}
